src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs

```diff
@@ -15,71 +16,125 @@ internal sealed partial class Checksum : IObjectWritable, IEquatable<Checksum>
+                result.Data3 = reader.ReadInt32();
+                return result;
             }
 
-            for (var i = 0; i < _checkSum.Length; i++)
+            public override int GetHashCode()
             {
-                if (_checkSum[i] != other._checkSum[i])
+                // The checksum is already a hash. Just read a 4-byte value to get a well-distributed hash code.
+                return (int)Data1;
+            }
+
+            public override bool Equals(object obj)
+            {
+                if (!(obj is Sha1Hash other))
                 {
                     return false;
                 }
+
+                return Equals(other);
+            }
+
+            public bool Equals(Sha1Hash other)
+            {
+                return Data1 == other.Data1
+                    && Data2 == other.Data2
+                    && Data3 == other.Data3;
             }
+        }
+
+        private Sha1Hash _checkSum;
 
-            return true;
+        public unsafe Checksum(byte[] checksum)
+        {
+            if (checksum.Length == 0)
+            {
+                _checkSum = default;
+                return;
+            }
+            else if (checksum.Length != sizeof(Sha1Hash))
+            {
+                throw new ArgumentException($"{nameof(checksum)} must be a SHA-1 hash", nameof(checksum));
+            }
+
+            fixed (byte* data = checksum)
+            {
+                _checkSum = *(Sha1Hash*)data;
+            }
         }
 
-        public override bool Equals(object obj)
+        private Checksum(Sha1Hash hash)
         {
-            return Equals(obj as Checksum);
+            _checkSum = hash;
         }
 
-        public override int GetHashCode()
+        public bool Equals(Checksum other)
         {
-            if (_lazyHash == 0)
+            if (other == null)
             {
-                _lazyHash = CalculateHashCode();
+                return false;
             }
 
-            return _lazyHash;
+            return _checkSum == other._checkSum;
         }
 
-        public override string ToString()
+        public override bool Equals(object obj)
         {
-            return Convert.ToBase64String(_checkSum);
+            return Equals(obj as Checksum);
         }
 
-        private int CalculateHashCode()
+        public override int GetHashCode()
         {
-            // lazily calculate hash for checksum
-            var hash = _checkSum.Length;
+            return _checkSum.GetHashCode();
+        }
 
-            for (var i = 0; i < _checkSum.Length; i++)
+        public override unsafe string ToString()
+        {
+            var data = new byte[sizeof(Sha1Hash)];
+            fixed (byte* dataPtr = data)
             {
-                hash = Hash.Combine((int)_checkSum[i], hash);
+                *(Sha1Hash*)dataPtr = _checkSum;
             }
 
-            // make sure we never return 0
-            return hash == 0 ? 1 : hash;
```

Comment:
➡️ The value 0 was only used to detect cases where the hash code (for `GetHashCode()`) was not initialized. Since we no longer use lazy initialization for the hash code, we no longer need this condition.